File: content/showcase/benchmarks/runtime/byte-scan/zig/byte_scan.zig
const std = @import("std");

const size: usize = 256 * 1024 * 1024;
const repeat: usize = 8;
const newline_stride: usize = 4093;
const delim_stride: usize = 1_000_003;

const Rng = struct {
    state: u64,
    fn next(self: *Rng) u64 {
        var x = self.state;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.state = x;
        return x;
    }
};

fn fill(words: []u64, buf: []u8) void {
    var rng = Rng{ .state = 0x2545F4914F6CDD1D };
    for (words) |*w| w.* = (rng.next() & 0x1f1f1f1f1f1f1f1f) | 0x4040404040404040;
    var i: usize = newline_stride;
    while (i < buf.len) : (i += newline_stride) buf[i] = '\n';
    const marks = [_]u8{ ',', ':', '"' };
    var which: usize = 0;
    var d: usize = delim_stride;
    while (d < buf.len) : (d += delim_stride) {
        buf[d] = marks[which];
        which = (which + 1) % 3;
    }
}

fn countNewlines(data: []const u8) i64 {
    var count: i64 = 0;
    var from: usize = 0;
    while (std.mem.indexOfScalarPos(u8, data, from, '\n')) |p| {
        count += 1;
        from = p + 1;
    }
    return count;
}

const Scan = struct { found: i64, index_sum: i64, which_sum: i64 };

fn scanDelims(data: []const u8) Scan {
    var r = Scan{ .found = 0, .index_sum = 0, .which_sum = 0 };
    var from: usize = 0;
    while (std.mem.indexOfAnyPos(u8, data, from, ",:\"")) |p| {
        r.found += 1;
        r.index_sum += @intCast(p);
        r.which_sum += switch (data[p]) {
            ',' => 0,
            ':' => 1,
            else => 2,
        };
        from = p + 1;
    }
    return r;
}

pub fn main() !void {
    const words = try std.heap.page_allocator.alloc(u64, size / 8);
    defer std.heap.page_allocator.free(words);
    const buf = std.mem.sliceAsBytes(words);
    fill(words, buf);
    var absent: i64 = 0;
    var newlines: i64 = 0;
    var delims: i64 = 0;
    var index_sum: i64 = 0;
    var which_sum: i64 = 0;
    var r: usize = 0;
    while (r < repeat) : (r += 1) {
        if (std.mem.indexOfScalar(u8, buf, 0) != null) absent += 1;
        newlines += countNewlines(buf);
        const s = scanDelims(buf);
        delims += s.found;
        index_sum += s.index_sum;
        which_sum += s.which_sum;
    }
    try std.io.getStdOut().writer().print("absent: {d} newlines: {d} delims: {d} index-sum: {d} which-sum: {d}\n", .{ absent, newlines, delims, index_sum, which_sum });
}

File: content/showcase/benchmarks/runtime/byte-scan/rust/byte_scan.rs
const SIZE: usize = 256 * 1024 * 1024;
const REPEAT: usize = 8;
const NEWLINE_STRIDE: usize = 4093;
const DELIM_STRIDE: usize = 1_000_003;

struct Rng(u64);

impl Rng {
    fn next(&mut self) -> u64 {
        let mut x = self.0;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.0 = x;
        x
    }
}

fn fill(buf: &mut [u8]) {
    let mut rng = Rng(0x2545F4914F6CDD1D);
    for word in buf.chunks_exact_mut(8) {
        word.copy_from_slice(&((rng.next() & 0x1f1f1f1f1f1f1f1f) | 0x4040404040404040).to_le_bytes());
    }
    let mut i = NEWLINE_STRIDE;
    while i < buf.len() {
        buf[i] = b'\n';
        i += NEWLINE_STRIDE;
    }
    let marks = [b',', b':', b'"'];
    let mut d = DELIM_STRIDE;
    let mut which = 0;
    while d < buf.len() {
        buf[d] = marks[which];
        which = (which + 1) % 3;
        d += DELIM_STRIDE;
    }
}

fn count_newlines(data: &[u8]) -> i64 {
    let mut count = 0;
    let mut from = 0;
    while let Some(p) = data[from..].iter().position(|&b| b == b'\n') {
        count += 1;
        from += p + 1;
    }
    count
}

fn scan_delims(data: &[u8]) -> (i64, i64, i64) {
    let (mut found, mut index_sum, mut which_sum) = (0, 0, 0);
    let mut from = 0;
    while let Some(p) = data[from..].iter().position(|&b| matches!(b, b',' | b':' | b'"')) {
        let index = from + p;
        found += 1;
        index_sum += index as i64;
        which_sum += match data[index] {
            b',' => 0,
            b':' => 1,
            _ => 2,
        };
        from = index + 1;
    }
    (found, index_sum, which_sum)
}

fn main() {
    let mut buf = vec![0u8; SIZE];
    fill(&mut buf);
    let (mut absent, mut newlines, mut delims, mut index_sum, mut which_sum) = (0, 0, 0, 0, 0);
    for _ in 0..REPEAT {
        if buf.contains(&0) {
            absent += 1;
        }
        newlines += count_newlines(&buf);
        let (f, i, w) = scan_delims(&buf);
        delims += f;
        index_sum += i;
        which_sum += w;
    }
    println!("absent: {} newlines: {} delims: {} index-sum: {} which-sum: {}", absent, newlines, delims, index_sum, which_sum);
}
